explorer/explorer.go

```diff
@@ -16,6 +22,50 @@ var (
 	ErrNotAvailable = errors.New("current OS not supported")
 )
 
+type result struct {
+	file  interface{}
+	error error
+}
+
+type Explorer struct {
+	id    int32
+	mutex sync.Mutex
+
+	// explorer holds OS-Specific content, it varies for each OS.
+	*explorer
+}
+
+// active holds all explorer currently active, that may necessary for callback functions.
+//
+// Some OSes (Android, iOS, macOS) may call Golang exported functions as callback, but we need
+// someway to link that callback with the respective explorer, in order to give them a response.
+//
+// In that case, a construction like `callback(..., id int32)` is used. Then, it's possible to get the explorer
+// by lookup the active using the callback id.
+//
+// To avoid hold dead/unnecessary explorer, the active will be removed using `runtime.SetFinalizer` on the related
+// Explorer.
+var active = sync.Map{} // map[int32]*explorer
+
+// NewExplorer creates a new Explorer for the given *app.Window.
+// The given app.Window must be unique and you should call NewExplorer
+// once per new app.Window.
+//
+// It's mandatory to use Explorer.ListenEvents on the same *app.Window.
+func NewExplorer(w *app.Window) (e *Explorer) {
+	idb, _ := rand.Int(rand.Reader, big.NewInt(math.MaxInt32))
+
+	e = &Explorer{
+		explorer: newExplorer(w),
+		id:       int32(idb.Int64()),
+	}
+
+	active.Store(e.id, e.explorer)
+	runtime.SetFinalizer(e, func(e *Explorer) { active.Delete(e.id) })
```

Comment:
I don't understand why we need to use a cryptographically random integer as the ID; can you explain it for me? I would have thought that an unexported package global counter (incremented using sync/atomic) would achieve the uniqueness that we need. Am I missing something?

As an aside, could we use the window pointer (converted to an int64 or something) as the ID? That would be unique among windows, would allow us to trivially test whether there is already an explorer for a given window, and would even allow `NewExplorer` to return the existing explorer for your window if it already exists. What do you think? I suspect I'm not seeing all of the tradeoffs of such a design.

Reply:
> I don't understand why we need to use a cryptographically random integer as the ID; can you explain it for me? I would have thought that an unexported package global counter (incremented using sync/atomic) would achieve the uniqueness that we need. Am I missing something?

I think `sync/atomic` would work. It just need a unique id, I use random because it's highly possible to be unique. 😂

> As an aside, could we use the window pointer (converted to an int64 or something) as the ID? That would be unique among windows, would allow us to trivially test whether there is already an explorer for a given window, and would even allow NewExplorer to return the existing explorer for your window if it already exists. What do you think? I suspect I'm not seeing all of the tradeoffs of such a design.

I thought about using the window pointer as id, but I'm afraid that the Golang reuses the exactly same pointer for some new window (after some GC). _I don't if it's possible (or likely to happen)._

Reply:
I think we're safe from pointer reuse because:

- the explorer has a pointer to the window, so it will always GC _before_ the window (this assumes that ever platform retains the window pointer, which isn't currently true).
- when the explorer GCs, the finalizer will remove the window address from the map before it can possibly be reused by another window

However, the above does point out that we need to stress the importance of destroying the explorer when you close the window. Otherwise, the explorer will be retained in the map and its reference to the window will cause the window also to escape garbage collection.

What if we convert the `sync.Map` to a package-scoped `map[*app.Window]*explorer` protected by a mutex? This ensures that every platform "retains" the window pointer even if it isn't a field in the explorer. The ID field can then by `uint64(uintptr(unsafe.Pointer(w)))`. It will still be the case that users must be careful to close the explorer, but that seems inevitable.

Reply:
I tried to change to `uint64(uintptr(unsafe.Pointer(w)))`, but it fails on Android, because on Android it uses `int` (not `long`). The `requestCode` is `int`. I think that is why I choose random ids.

I think using `int32(uintptr(unsafe.Pointer(w)))` can lead to more collisions. I tought about using some `fnv` to hash it, but it's as bad as random ids. something like `h := fnv.New32(); h.Write((*[unsafe.Sizeof(new(int))]byte)(unsafe.Pointer(&id))[:]);`.

Reply:
Let's switch to an atomic counter then.

Reply:
I fix that on `explorer: replace random id by atomic counter`.